core/backend/src/StashManager.ts

```diff
@@ -0,0 +1,458 @@
+    const shared = this.queryLocks(args, LockState.Shared, LockOrigin.Acquired);
+    await args.db.locks.acquireLocks({ shared });
+
+    const exclusive = this.queryLocks(args, LockState.Exclusive, LockOrigin.Acquired);
+    await args.db.locks.acquireLocks({ exclusive });
+
+    const newElements = this.queryLocks(args, LockState.Shared, LockOrigin.NewElement);
+    for (const id of newElements) {
+      if (!args.db.locks.holdsExclusiveLock(id)) {
+        args.db.locks[_elementWasCreated](id);
+      }
+    }
+  }
+  /**
+   * Creates a stash of changes for the specified briefcase.
+   *
+   * This method generates a stash in the stash root directory for the given briefcase, using the provided description and iModelId.
+   * Optionally, it can reset the briefcase by releasing all locks after stashing.
+   *
+   * @param args - The properties required to create a stash, including the briefcase, description, iModelId, and an optional resetBriefcase flag.
+   * @returns A promise that resolves to the properties of the created stash.
+   *
+   * @throws Error if the stash operation fails.
+   */
+  public static async stash(args: CreateStashProps): Promise<StashProps> {
+    const stashRootDir = StashManager.getStashRootFolder(args.db, true);
+    const iModelId = args.db.iModelId;
+    if (!args.db.txns.hasPendingTxns) {
+      throw new StashError("nothing to stash");
+    }
+
+    if (args.db.txns.hasUnsavedChanges) {
+      throw new StashError("Unsaved changes exist");
+    }
+
+    if (args.db.txns.hasPendingSchemaChanges) {
+      throw new StashError("Pending schema changeset. Stashing is not currently supported for schema changes");
+    }
+
+    const stash = args.db[_nativeDb].stashChanges({ stashRootDir, description: args.description, iModelId }) as StashProps;
+    if (args.discardLocalChanges) {
+      await args.db.discardChanges({ retainLocks: args.retainLocks });
+    }
+
+    Logger.logInfo(loggerCategory, `Stashed changes`, () => stash);
+    return stash;
+  }
+
+  /**
+   * Retrieves the stash properties from the database for the given arguments.
+   *
+   * This method attempts to read the stash information from the local database using a prepared SQL statement.
+   * If successful, it returns the stash properties along with the file path of the stash database.
+   * If an error occurs during the process, it logs the error and returns `undefined`.
+   *
+   * @param args - The arguments required to locate and access the stash.
+   * @returns The stash file properties if found; otherwise, `undefined`.
+   */
+  public static getStash(args: StashArgs): StashProps | undefined {
+    try {
+      return this.withStash(args, (stashDb) => {
+        const stashProps = stashDb.withPreparedSqliteStatement("SELECT [val] FROM [be_Local] WHERE [name]='$stash_info'", (stmt) => {
+          if (stmt.step() !== DbResult.BE_SQLITE_ROW)
+            throw new IModelError(IModelStatus.BadArg, "Invalid stash");
```

Comment:
Why do we use `IModelError` here and below and `StashError` in almost everywhere for the rest of the file?

Reply:
all IModelError removed from stashManager and replaced with StashError.